Add fetchMode modline

RobinHoutevelts opened the pull request:
With this PR you'll be able to change the used PDO fetch mode on your select queries.

There is support for `assoc`, `class`, `column`, `keypair` and `named`.

```sql
-- name: getNormal
select id, something from test_table

-- name: getKeyed fetchMode: keypair
select id, something from test_table
```
```php
// Key using a foreach
$keyed = [];
foreach($r->getNormal() as $row) {
  $keyed[$row['id']] = $row['something'];
}

// Or simply
$keyed = $r->getKeyed();
```

The fetch modes `class` and `column` require arguments ( classname and column number ). You can pass those in parentheses. The `column` fetchmode has a default argument of `0`.

Eg:

```sql
-- name: getObjectById oneOrMany: one fetchMode: class(MyObject)
select * from test_table where id = ?
-- name: getColumn fetchMode: column
select id from test_table
```
```php
// returns one row, which is an instance of MyObject with id and something set
class MyObject {
}
$r->getObjectById(3);
// returns a one-dimensional array of ids
$r->getColumn();
```

RobinHoutevelts committed:
Add fetchMode modline

RobinHoutevelts committed:
Fix test

RobinHoutevelts committed:
sssh, it's kk bby

nulpunkt commented:
Hi, thanks a lot for taking the time to do this!

It's a pretty cool idea to be able to get key value pairs like this, but being able to just specify any fetch mode feels like opening a can of worms. The <http://php.net/manual/en/pdo.constants.php#pdo.constants.fetch-group> would not be fully supported with your implementation and I really don't see a nice way of supporting it.

I'm thinking that we might want to simply support the most useful fetch modes, like `FETCH_KEY_PAIR`, `FETCH_NAMED` and `FETCH_COLUMN` as explicit fetch modes, so you specify: `fetchMode: (keypair|named|column)` and we manually map those to appropriate fetch modes.

What are your thoughts?

Ps:
When sending pull requests it would be nice if you used `git commit --amend` to commit changes so we don't get a confusing history. You can also use `git rebase -i` to squash your commits.

terales commented:
> I'm thinking that we might want to simply support the most useful fetch modes, like FETCH_KEY_PAIR, FETCH_NAMED and FETCH_COLUMN as explicit fetch modes, so you specify: fetchMode: (keypair|named|column) and we manually map those to appropriate fetch modes.

+1 for this idea. I've never seen usage of other fetch modes in commercial projects

RobinHoutevelts commented:
I'll update the PR later today :+1:

RobinHoutevelts committed:
Limit and refactor fetchmodes

RobinHoutevelts committed:
Fix tests

RobinHoutevelts committed:
Have the fetchmodeFactory be protected

RobinHoutevelts committed:
Awkward

RobinHoutevelts committed:
Fix typo

RobinHoutevelts committed:
Don't limit the fetchmodes in regex

This way when a user makes a typo we can show an error.

Using \w+ instead of \S+ because I don't want to match parentheses.

RobinHoutevelts commented:
Sorry for the delay. Had a very busy week.

You were right that specifying any fetch mode from the modline without having any way to massage it from code was a bad thing to do 🙊

For example:

```sql
-- name: getColumn fetchMode: COLUMN
select id from test_table
```

Failed because `PDO::FETCH_COLUMN` required a column number. It didn't fallback to `0` as I hoped.

Sorry about that.

---

I've added support for the fetchmodes you posted. And made sure that adding more in the future can be done.

I've **deprecated** the `rowClass: MyObject ` mod. Since it's basically `fetchMode: class(MyObject)`.

I've replaced the `rowClass` example in the README with a `fetchMode: class` example.

[FetchMode/Factory](https://github.com/RobinHoutevelts/yesql-php/blob/6ca1e36ed5d2cc0012d2dc25cd8beff92a926e54/src/Nulpunkt/Yesql/FetchMode/Factory.php#L40) **still handles** `rowClass` so you don't have to go and update all your existing projects.

---

Adding a new Fetchmode should be easy. Just extend [FetchMode/BaseFetchMode](https://github.com/RobinHoutevelts/yesql-php/blob/6ca1e36ed5d2cc0012d2dc25cd8beff92a926e54/src/Nulpunkt/Yesql/FetchMode/BaseFetchMode.php) and implement the `setMode()` method.

In the [FetchMode/Factory](https://github.com/RobinHoutevelts/yesql-php/blob/6ca1e36ed5d2cc0012d2dc25cd8beff92a926e54/src/Nulpunkt/Yesql/FetchMode/Factory.php#L13) you have to map your desired fetchmode name to your new class.

For example to add a fetchmode that groups by the first column:

```
-- name: getGroupedBySex fetchMode: my_grouping_fetch_mode
select sex, something from test_table
```

```php
//FetchMode/Factory.php
public function createFromModLine($modLine)
{
    ...

    $fetchModes = [
        ...
        'my_grouping_fetch_mode' => MyNewFetchMode::class,
    ];

    ...
```

 ```php
//FetchMode/MyNewFetchMode.php
class MyNewFetchMode extends BaseFetchMode
{
    public function setMode(PDOStatement $stmt)
    {
        $stmt->setFetchMode(
            \PDO::FETCH_GROUP | \PDO::FETCH_COLUMN
        );
    }
}
```


----

It would be cool if users could register their own fetch modes. But that's a bit out of scope for this PR.

```php
$repository = new Repository($db, $file);
$repository->addFetchMode('foobar', MyNewFetchMode::class);
```

RobinHoutevelts commented:
> Ps:
When sending pull requests it would be nice if you used git commit --amend to commit changes so we don't get a confusing history. You can also use git rebase -i to squash your commits.

Actually during a PR I'd like to avoid that since we would no longer see the individual changes during code review.

As soon as you give the OK I'll squash and commit so your history isn't garbage 😄

nulpunkt commented:
No problem man, I'm pretty busy at work, so sorry for my delays as well!

I really like how this is turning out! Super useful stuff!

Could you please add tests for all the modlines?

nulpunkt commented:
I feel like this can be considered dead now. I'd love to see this pr return :)
